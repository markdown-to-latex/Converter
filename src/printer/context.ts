import { getLatexApplicationLetter } from './latex';

export type WriteFileFunction = (
    content: string,
    fileName: string,
    context: Context,
) => void;

export interface Context {
    config: {
        defaultFontSize: 14;
    };
    applications: {
        accessKeys: string[];
        keyToData: Record<
            string,
            {
                title: string;
                text: (label: string) => string;
            }
        >;
    };
    references: {
        accessKeys: string[];
        keyToData: Record<
            string,
            {
                text: (label: string) => string;
            }
        >;
    };
    picture: {
        key: string;
        label: string;
        height: string;
        keyToLabel: Record<string, string>;
    };
    table: {
        key: string;
        label: string;
        keyToLabel: Record<string, string>;
    };
    code: {
        key: string;
        label: string;
        lang: string;
    };
    writeFile: WriteFileFunction;
}

export function getOrCreatePictureLabel(context: Context, key: string): string {
    let label = context.picture.keyToLabel[key];
    if (label !== undefined) {
        return label;
    }

    const lastLabel = Math.max(
        0,
        ...Object.values(context.picture.keyToLabel).map(v => +v),
    );
    label = (lastLabel + 1).toString();
    context.picture.keyToLabel[key] = label;
    return label;
}

export function getOrCreateTableLabel(context: Context, key: string): string {
    let label = context.table.keyToLabel[key];
    if (label !== undefined) {
        return label;
    }

    const lastLabel = Math.max(
        0,
        ...Object.values(context.table.keyToLabel).map(v => +v),
    );
    label = (lastLabel + 1).toString();
    context.table.keyToLabel[key] = label;
    return label;
}

class ContextError extends Error {}

export function getApplicationLabelByKey(
    context: Context,
    key: string,
): string {
    if (context.applications.keyToData[key] === undefined) {
        throw new ContextError(`Application with key "${key}" not found`);
    }

    let index = context.applications.accessKeys.indexOf(key);
    if (index === -1) {
        index = context.applications.accessKeys.length;
        context.applications.accessKeys.push(key);
    }

    return getLatexApplicationLetter(index);
}

export function addApplicationByKey(
    context: Context,
    key: string,
    data: Context['applications']['keyToData'][0],
): void {
    if (context.applications.keyToData[key] !== undefined) {
        throw new ContextError(`Application with key "${key}" already exists`);
    }

    context.applications.keyToData[key] = data;
}

export function getReferenceLabelByKey(context: Context, key: string): string {
    if (context.references.keyToData[key] === undefined) {
        throw new ContextError(`Reference with key "${key}" not found`);
    }

    let index = context.references.accessKeys.indexOf(key);
    if (index === -1) {
        index = context.references.accessKeys.length;
        context.references.accessKeys.push(key);
    }

    return index.toString();
}

export function addReferenceByKey(
    context: Context,
    key: string,
    data: Context['references']['keyToData'][0],
): void {
    if (context.references.keyToData[key] !== undefined) {
        throw new ContextError(`Reference with key "${key}" already exists`);
    }

    context.references.keyToData[key] = data;
}
