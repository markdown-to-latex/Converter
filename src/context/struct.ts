import { Node } from '../ast/node';
import { DiagnoseList } from '../diagnose';
import { ContextConfig } from '../printer/context';

export interface ContextApplicationInfo {
    title: Node[];
    label: string;
}

export interface ContextApplicationContentInfo extends ContextApplicationInfo {
    content: (this: ContextApplicationContentInfo, indexText: Node[]) => Node[];
}

export interface ContextReferenceInfo {
    label: string;
}

export interface ContextReferenceContentInfo extends ContextReferenceInfo {
    content: (this: ContextReferenceContentInfo, indexText: Node[]) => Node[];
}

export interface ContextPictureInfo {
    label: string;
}

export interface ContextTableInfo {
    label: string;
}

export interface ContextInfoReference {
    refs: number;
    node: Node;
}

export interface ContextTemp {
    application: ContextApplicationInfo | null;
    reference: ContextReferenceInfo | null;

    /**
     * Current node
     */
    node: Node;
}

export interface ContextDataGeneric<T> {
    /**
     * Relocation (indexes post-processing)
     */
    labels: string[];

    /**
     * Labels to node mapping (for diagnostic positioning)
     */
    labelToRefs: Record<string, ContextInfoReference>

    /**
     * Mapping labels to object information
     * Fills only if the target object has been defined
     */
    labelToInfo: Record<string, T>;
}

export type ContextApplicationData =
    ContextDataGeneric<ContextApplicationContentInfo>;
export type ContextReferenceData =
    ContextDataGeneric<ContextReferenceContentInfo>;
export type ContextPictureData = ContextDataGeneric<ContextPictureInfo>;
export type ContextTableData = ContextDataGeneric<ContextTableInfo>;

export interface ContextData {
    application: ContextApplicationData;
    reference: ContextReferenceData;
    picture: ContextPictureData;
    table: ContextTableData;
}

export interface Context {
    diagnostic: DiagnoseList;
    config: ContextConfig;

    data: ContextData;
    temp: ContextTemp;

    writeDiagnosticList: (this: Context) => void;
    writeFile: (this: Context, fileName: string, content: string) => void;
}
