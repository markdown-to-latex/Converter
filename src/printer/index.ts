import { Node } from '../ast/node';
import { Context } from './context';
import { applyPrinterVisitors } from './visitors';

export function printMarkdownAST(rootNode: Node, context: Context): void {
    context = context ?? {};

    applyPrinterVisitors(rootNode, context);
}
