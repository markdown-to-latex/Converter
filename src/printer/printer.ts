import { Node } from '../ast/nodes';
import { WriteFileFunction } from './context';
import { applyPrinterVisitors } from './visitors';

function prettifyLaTeX(text: string): string {
    return text.replace(/\n{3,}/g, '\n\n');
}

export function printMarkdownAST(
    rootNode: Node,
    writeFile: WriteFileFunction,
): void {
    applyPrinterVisitors(rootNode, {
        writeFile: (content, fileName, context) => {
            content = prettifyLaTeX(content);
            return writeFile(content, fileName, context);
        },
        code: {
            key: '',
            label: '',
            lang: '',
        },
        applications: {
            accessKeys: [],
            keyToData: {},
        },
        picture: {
            key: '',
            height: '',
            keyToLabel: {},
            label: '',
        },
        references: {
            accessKeys: [],
            keyToData: {},
        },
        table: {
            key: '',
            label: '',
            keyToLabel: {},
        },
        config: {
            defaultFontSize: 14,
        },
    });
}
