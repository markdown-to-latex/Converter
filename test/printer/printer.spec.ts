import {
    applyProcessing,
    buildMarkdownAST,
    initContext,
    lexer,
    printMarkdownAST,
} from '../../src';
import { OpCodeError } from '../../src/printer/opcodes';
import { ContextError } from '../../src/printer/context';
import { MarkDownToLaTeXConverter } from '../../src/printer/types';

function processingChain(
    text: string,
    config?: Partial<MarkDownToLaTeXConverter>,
): Record<string, string> {
    const lexerResult = lexer(text);
    const result = buildMarkdownAST(lexerResult, { filepath: 'filepath' })

    const files: Record<string, string> = {};
    const context = initContext((content, fileName) => {
        files[fileName] = content;
    }, config);

    applyProcessing(result, context);

    printMarkdownAST(result, context);

    return files;
}

describe('simple md to latex docs printer', () => {
    test('Paragraph', () => {
        const result = processingChain(`
# Header

Text
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Subheader + List + Code Span', () => {
        const result = processingChain(`
# Header

- A
- B
- C

## Subheader

1. X
2. Y
    1. T
        - 600
        - 700
    2. \`Code_span\`
3. Z
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Header + Image + Code + Image', () => {
        const result = processingChain(`
# Header

!P[img-1!5cm]
![Image name](./assets/img/dolphin.png)

!C[code-1!Python Sample Code]
\`\`\`python
def main():
    print "Hello World"
\`\`\`

!P[img-2!7cm]
![Image name 2](./assets/img/dolphin.png)
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Code + Code', () => {
        const result = processingChain(`
# Header

Code in !PK[code-1] Ð¸ !PK[code-2].

!C[code-1!Python Sample Code]
\`\`\`python
def main():
    print "Hello World"
\`\`\`

!C[code-2!Python Sample Code 2]
\`\`\`python
def hello_world():
    print "Hello World"
\`\`\`
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Table', () => {
        const result = processingChain(`
Demonstrated in table  

!T[table!Table with content]

|a|b|c|d|
|---|---|---|---|
|1|2|3|4|
|t|r|e|z|
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Header + Formula', () => {
        const result = processingChain(`
# Header

\`\`\`math
    a = b + c
\`\`\`
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('bold and italic', () => {
        const result = processingChain(`**Bold**: *testing*`)['filepath'];

        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });
});

describe('Applications', () => {
    test('with list', () => {
        const result = processingChain(`
!AC[code-full!./assets/code!template-full.py!python]
!AC[code-full2!./assets/code!template-full2.py!python]
!APR[picture-large!Large scheme!./assets/img/circuit.png]
        
# Header

Code from application !AK[code-full2] describes image from application !AK[picture-large].

See application !AK[code-full].

# Applications

!LAA[]
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('with multiple columns', () => {
        const result = processingChain(`
!ACC[2]
!AC[code-full!./assets/code!template-full.py!python]
        
# Header

See application !AK[code-full].

# Applications

!LAA[]
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Unused application, should throw error', () => {
        const result = () =>
            processingChain(`
!AC[code-full!./assets/code!template-full.py!python]

!LAA[]
`)['filepath'];
        expect(result).toThrow(OpCodeError);
    });

    test('Undefined application, should throw error', () => {
        const result = () =>
            processingChain(`
!AK[nope]
`)['filepath'];
        expect(result).toThrow(ContextError);
    });
});

describe('References', () => {
    test('with list', () => {
        const result = processingChain(`
!RR[ref-1]
\`\`\`ref
H.\\,Y.\\~Ignat. "Reference\\~1" // Some Journal, 1867
\`\`\`

!RR[ref-2]
\`\`\`ref
H.\\,Y.\\~Ignat. "Reference\\~2" // Some Journal, 1867
\`\`\`
        
# Header

Code from reference !RK[ref-2] describes image from reference !RK[ref-1].

# References

!LAR[]
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Unused reference, should throw error', () => {
        const result = () =>
            processingChain(`
!RR[ref]
\`\`\`ref
A.\\,A.\\~Amogus. "Impostor\\~theorem" // Steam library, 2021
\`\`\`

!LAR[]
`)['filepath'];
        expect(result).toThrow(OpCodeError);
    });

    test('Undefined reference, should throw error', () => {
        const result = () =>
            processingChain(`
!RK[nope]
`)['filepath'];
        expect(result).toThrow(ContextError);
    });

    test('Inline math', () => {
        const result = processingChain(`
Text $\`a = b + \\sum_{i=0}^\\infty c_i\`$ ending.
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result)
            .toEqual(`Text $\\displaystyle a = b + \\sum_{i=0}^\\infty c_i$ ending.
`);
    });

    test('Text with percents', () => {
        const result = processingChain(`
Text with 10% number.
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Text with escapes ("<" sound be corrent also)', () => {
        const result = processingChain(`
Text with \\<assdasd.
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Tag <hr> should break the page', () => {
        const result = processingChain(`
The first page

---------------------------------------

The second page
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Tag <br> should put additional break', () => {
        const result = processingChain(`
The first line  
The second line
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test("Text ' dereplacement", () => {
        const result = processingChain(`
Otsu's method is a one-dimensional discrete analog of Fisher's 
Discriminant Analysis, is related to Jenks optimization method.
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result)
            .toEqual(`Otsu's method is a one-dimensional discrete analog of Fisher's 
Discriminant Analysis, is related to Jenks optimization method.
`);
    });

    test('Inline latex math dereplacement', () => {
        const result = processingChain(`
$\`a > b < c\`$
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Table and picture key', () => {
        const result = processingChain(`
Displayed in picture !PK[gray-square] (!PK[gray-square]) and table !TK[table].

!P[gray-square!5cm]
![Gray square](./assets/img/example.png)

!T[table!Table]
        
|Key    |Value |
|-------|------|
|Static number | 50 |
|Random number | $$ \\showcaserandomnumber $$ |
`)['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });
});

describe('latex picture after table (#52)', function () {
    // See https://github.com/markdown-to-latex/converter/issues/52
    test('Picture right after the table', () => {
        const result = processingChain(
            `!T[table!Table example]

| Key           | Value                       |
| ------------- | --------------------------- |
| Static number | 50                          |

!P[gray-square!5cm]
![Gray square](./assets/img/example.png)`,
        )['filepath'];

        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Table + text + picture', () => {
        const result = processingChain(
            `!T[table!Table example]

| Key           | Value                       |
| ------------- | --------------------------- |
| Static number | 50                          |

Sample text line

!P[gray-square!5cm]
![Gray square](./assets/img/example.png)`,
        )['filepath'];

        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });
});

describe('url variants', () => {
    test('Default url', () => {
        const result = processingChain(
            'https://example.com/index_page.html?asd=asdasd&gege=gegege#header',
        )['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Bold url', () => {
        const result = processingChain(
            'https://example.com/index_page.html?asd=asdasd&gege=gegege#header',
            {
                latex: {
                    useLinkAs: 'bold',
                },
            },
        )['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Italic url', () => {
        const result = processingChain(
            'https://example.com/index_page.html?asd=asdasd&gege=gegege#header',
            {
                latex: {
                    useLinkAs: 'italic',
                },
            },
        )['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('Underlined url', () => {
        const result = processingChain(
            'https://example.com/index_page.html?asd=asdasd&gege=gegege#header',
            {
                latex: {
                    useLinkAs: 'underline',
                },
            },
        )['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });

    test('No escape & code url', () => {
        const result = processingChain(
            'https://example.com/index_page.html?asd=asdasd&asdasd=gege#header',
            {
                latex: {
                    useLinkAs: 'code',
                    defaultAutoEscapes: false,
                },
            },
        )['filepath'];
        expect(result).not.toBeUndefined();
        expect(result).toMatchSnapshot();
    });
});
