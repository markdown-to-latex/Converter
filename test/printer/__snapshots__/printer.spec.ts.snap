// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applications with list 1`] = `
"\\\\subtitle{Header}

Code from application А describes image from application Б.

See application В.

\\\\subtitle{Applications}

\\\\pagebreak
\\\\subtitle{Приложение А}

\\\\section*{Листинг кода из файла template-full2.py}

\\\\vspace{1em}
\\\\fontsize{\\\\applicationcodefontsize}{\\\\applicationcodefontsize}\\\\selectfont

\\\\inputminted[baselinestretch=\\\\applicationcodelineheight]{python}{./assets/code/template-full2.py}

\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont

\\\\pagebreak
\\\\begin{landscape}
    \\\\thispagestyle{empty}
    \\\\subtitle{Приложение Б}

    \\\\section*{Large scheme}
    
    \\\\vspace{1em}
    \\\\begin{center}
    \\\\includegraphics[height=13.8cm]{./assets/img/circuit.png}
    \\\\end{center}

    \\\\vfill
    \\\\raisebox{.6ex}{\\\\makebox[\\\\linewidth]{\\\\thepage}}
\\\\end{landscape}

\\\\pagebreak
\\\\subtitle{Приложение В}

\\\\section*{Листинг кода из файла template-full.py}

\\\\vspace{1em}
\\\\fontsize{\\\\applicationcodefontsize}{\\\\applicationcodefontsize}\\\\selectfont

\\\\inputminted[baselinestretch=\\\\applicationcodelineheight]{python}{./assets/code/template-full.py}

\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont
"
`;

exports[`Applications with multiple columns 1`] = `
"\\\\subtitle{Header}

See application А.

\\\\subtitle{Applications}

\\\\pagebreak
\\\\subtitle{Приложение А}

\\\\section*{Листинг кода из файла template-full.py}

\\\\vspace{1em}
\\\\fontsize{\\\\applicationcodefontsize}{\\\\applicationcodefontsize}\\\\selectfont
\\\\begin{multicols}{2}
\\\\inputminted[baselinestretch=\\\\applicationcodelineheight]{python}{./assets/code/template-full.py}
\\\\end{multicols}
\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont
"
`;

exports[`CodeSpan Monospace 1`] = `
"CodeSpan \\\\texttt{text \\\\& text}.
"
`;

exports[`CodeSpan Quote 1`] = `
"CodeSpan <<text \\\\& text>>.
"
`;

exports[`Escapes Default escapes 1`] = `
"\\\\subtitle{Header}

\\\\underline{https://testing.url/com?some=thing\\\\&wtf\\\\#xdxdxd}

The \\"{}definition\\"{} increased by 1\\\\% (more text more text).
    
"
`;

exports[`References CodeSpan dereplacement 1`] = `
"\\\\texttt{\\"{}sample \\\\& text\\"{}}
"
`;

exports[`References Inline latex math dereplacement 1`] = `
"$\\\\displaystyle a > b < c$
"
`;

exports[`References Table and picture key 1`] = `
"Displayed in picture 1 (1) and table 1.

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\addtolength{\\\\belowcaptionskip}{-1.6em}
\\\\setlength{\\\\abovecaptionskip}{0.5em}

\\\\begin{figure}[H]
    \\\\centering
    \\\\includegraphics[height=5cm]{./assets/img/example.png}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty,margin={0pt,0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 1 -- Gray square}
\\\\end{figure}

\\\\fontsize{\\\\tablefontsize}{\\\\tablefontsize}\\\\selectfont
\\\\setlength{\\\\belowcaptionskip}{0em}
\\\\setlength{\\\\abovecaptionskip}{0em}
\\\\setlength{\\\\LTpre}{2em}
\\\\setlength{\\\\LTpost}{2em}

\\\\begin{longtable}[H]{|c|c|}
    \\\\captionsetup{justification=justified,indention=0cm,labelformat=empty, margin={2pt, 0cm},font={stretch=1.5}}
    \\\\caption{Таблица 1 -- Table}
    \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endfirsthead
    \\\\caption{Продолжение таблицы 1} \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endhead
    \\\\endfoot
    \\\\endlastfoot

Static number & 50\\\\\\\\ \\\\hline
Random number &  \\\\showcaserandomnumber \\\\\\\\ \\\\hline

\\\\end{longtable}
\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont\\\\setstretch{1.5}
"
`;

exports[`References Tag <br> should put additional break 1`] = `
"The first line

The second line
"
`;

exports[`References Tag <hr> should break the page 1`] = `
"The first page

\\\\pagebreak
The second page
"
`;

exports[`References Text ' dereplacement 1`] = `
"Otsu's method is a one-dimensional discrete analog of Fisher's 
Discriminant Analysis, is related to Jenks optimization method.
"
`;

exports[`References Text with escapes ("<" sound be corrent also) 1`] = `
"Text with \\\\<assdasd.
"
`;

exports[`References Text with percents 1`] = `
"Text with 10\\\\% number.
"
`;

exports[`References with list 1`] = `
"\\\\subtitle{Header}

Code from reference 1 describes image from reference 2.

\\\\subtitle{References}

1.\\\\,H.\\\\,Y.\\\\~Ignat. \\"Reference\\\\~2\\" // Some Journal, 1867

2.\\\\,H.\\\\,Y.\\\\~Ignat. \\"Reference\\\\~1\\" // Some Journal, 1867
"
`;

exports[`latex picture after table (#52) Picture right after the table 1`] = `
"\\\\fontsize{\\\\tablefontsize}{\\\\tablefontsize}\\\\selectfont
\\\\setlength{\\\\belowcaptionskip}{0em}
\\\\setlength{\\\\abovecaptionskip}{0em}
\\\\setlength{\\\\LTpre}{2em}
\\\\setlength{\\\\LTpost}{0em}

\\\\begin{longtable}[H]{|c|c|}
    \\\\captionsetup{justification=justified,indention=0cm,labelformat=empty, margin={2pt, 0cm},font={stretch=1.5}}
    \\\\caption{Таблица 1 -- Table example}
    \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endfirsthead
    \\\\caption{Продолжение таблицы 1} \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endhead
    \\\\endfoot
    \\\\endlastfoot

Static number & 50\\\\\\\\ \\\\hline

\\\\end{longtable}
\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont\\\\setstretch{1.5}

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\setlength{\\\\abovecaptionskip}{0.5em}

\\\\begin{figure}[H]
    \\\\centering
    \\\\includegraphics[height=5cm]{./assets/img/example.png}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty,margin={0pt,0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 1 -- Gray square}
\\\\end{figure}
"
`;

exports[`latex picture after table (#52) Table + text + picture 1`] = `
"\\\\fontsize{\\\\tablefontsize}{\\\\tablefontsize}\\\\selectfont
\\\\setlength{\\\\belowcaptionskip}{0em}
\\\\setlength{\\\\abovecaptionskip}{0em}
\\\\setlength{\\\\LTpre}{2em}
\\\\setlength{\\\\LTpost}{2em}

\\\\begin{longtable}[H]{|c|c|}
    \\\\captionsetup{justification=justified,indention=0cm,labelformat=empty, margin={2pt, 0cm},font={stretch=1.5}}
    \\\\caption{Таблица 1 -- Table example}
    \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endfirsthead
    \\\\caption{Продолжение таблицы 1} \\\\\\\\\\\\hline
    Key & Value\\\\\\\\ \\\\hline

    \\\\endhead
    \\\\endfoot
    \\\\endlastfoot

Static number & 50\\\\\\\\ \\\\hline

\\\\end{longtable}
\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont\\\\setstretch{1.5}

Sample text line

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\setlength{\\\\abovecaptionskip}{0.5em}

\\\\begin{figure}[H]
    \\\\centering
    \\\\includegraphics[height=5cm]{./assets/img/example.png}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty,margin={0pt,0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 1 -- Gray square}
\\\\end{figure}
"
`;

exports[`simple md to latex docs printer Code + Code 1`] = `
"\\\\subtitle{Header}

Code in 1 и 2.

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\addtolength{\\\\belowcaptionskip}{-1.6em}
\\\\setlength{\\\\abovecaptionskip}{-0.5em}

\\\\begin{figure}[H]
    \\\\fontsize{\\\\codefontsize}{\\\\codefontsize}\\\\selectfont
    \\\\begin{minted}
    [baselinestretch=1.2]{python}
def main():
    print \\"Hello World\\"
    \\\\end{minted}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty, margin={0pt, 0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 1 -- Python Sample Code}
\\\\end{figure}

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\setlength{\\\\abovecaptionskip}{-0.5em}

\\\\begin{figure}[H]
    \\\\fontsize{\\\\codefontsize}{\\\\codefontsize}\\\\selectfont
    \\\\begin{minted}
    [baselinestretch=1.2]{python}
def hello_world():
    print \\"Hello World\\"
    \\\\end{minted}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty, margin={0pt, 0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 2 -- Python Sample Code 2}
\\\\end{figure}
"
`;

exports[`simple md to latex docs printer Del node 1`] = `
"Test~node \\\\textit{what} hell~yeah we~ll.
        
"
`;

exports[`simple md to latex docs printer Header + Formula 1`] = `
"\\\\subtitle{Header}

\\\\setlength{\\\\abovedisplayskip}{-0.9em}
\\\\setlength{\\\\belowdisplayskip}{0pt}
\\\\setlength{\\\\abovedisplayshortskip}{0pt}
\\\\setlength{\\\\belowdisplayshortskip}{0pt}
\\\\begin{align*}
\\\\displaystyle
    a = b + c
\\\\end{align*}    
"
`;

exports[`simple md to latex docs printer Header + Image + Code + Image 1`] = `
"\\\\subtitle{Header}

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\addtolength{\\\\belowcaptionskip}{-1.6em}
\\\\setlength{\\\\abovecaptionskip}{0.5em}

\\\\begin{figure}[H]
    \\\\centering
    \\\\includegraphics[height=5cm]{./assets/img/dolphin.png}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty,margin={0pt,0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 1 -- Image name}
\\\\end{figure}

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\addtolength{\\\\belowcaptionskip}{-1.6em}
\\\\setlength{\\\\abovecaptionskip}{-0.5em}

\\\\begin{figure}[H]
    \\\\fontsize{\\\\codefontsize}{\\\\codefontsize}\\\\selectfont
    \\\\begin{minted}
    [baselinestretch=1.2]{python}
def main():
    print \\"Hello World\\"
    \\\\end{minted}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty, margin={0pt, 0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 2 -- Python Sample Code}
\\\\end{figure}

\\\\setlength{\\\\intextsep}{3em}
\\\\setlength{\\\\belowcaptionskip}{-4ex}
\\\\setlength{\\\\abovecaptionskip}{0.5em}

\\\\begin{figure}[H]
    \\\\centering
    \\\\includegraphics[height=7cm]{./assets/img/dolphin.png}
    \\\\captionsetup{justification=centering,indention=0cm,labelformat=empty,margin={0pt,0cm},font={stretch=1.5}}
    \\\\caption{Рисунок 3 -- Image name 2}
\\\\end{figure}
"
`;

exports[`simple md to latex docs printer ListItem + Br + Text must have 2 line breaks 1`] = `
"\\\\hspace{0cm}а)\\\\,Item

New Line

\\\\hspace{0cm}б)\\\\,New Item
"
`;

exports[`simple md to latex docs printer ListItem + MathLatex must have 2 line breaks 1`] = `
"\\\\hspace{0cm}а)\\\\,Text

\\\\setlength{\\\\abovedisplayskip}{-0.9em}
\\\\setlength{\\\\belowdisplayskip}{0pt}
\\\\setlength{\\\\abovedisplayshortskip}{0pt}
\\\\setlength{\\\\belowdisplayshortskip}{0pt}
\\\\begin{align*}
\\\\displaystyle
Some text here
\\\\end{align*}    
"
`;

exports[`simple md to latex docs printer Paragraph 1`] = `
"\\\\subtitle{Header}

Text
"
`;

exports[`simple md to latex docs printer Subheader + List + Code Span 1`] = `
"\\\\subtitle{Header}

\\\\hspace{0cm}-\\\\,A

\\\\hspace{0cm}-\\\\,B

\\\\hspace{0cm}-\\\\,C

\\\\section{Subheader}

\\\\hspace{0cm}а)\\\\,X

\\\\hspace{0cm}б)\\\\,Y

\\\\hspace{1.25cm}1)\\\\,T

\\\\hspace{2.5cm}-\\\\,600

\\\\hspace{2.5cm}-\\\\,700

\\\\hspace{1.25cm}2)\\\\,\\\\texttt{Code\\\\_span}

\\\\hspace{0cm}в)\\\\,Z
"
`;

exports[`simple md to latex docs printer Table 1`] = `
"Demonstrated in table  

\\\\fontsize{\\\\tablefontsize}{\\\\tablefontsize}\\\\selectfont
\\\\setlength{\\\\belowcaptionskip}{0em}
\\\\setlength{\\\\abovecaptionskip}{0em}
\\\\setlength{\\\\LTpre}{2em}
\\\\setlength{\\\\LTpost}{2em}

\\\\begin{longtable}[H]{|c|c|c|c|}
    \\\\captionsetup{justification=justified,indention=0cm,labelformat=empty, margin={2pt, 0cm},font={stretch=1.5}}
    \\\\caption{Таблица 1 -- Table with content}
    \\\\\\\\\\\\hline
    a & b & c & d\\\\\\\\ \\\\hline

    \\\\endfirsthead
    \\\\caption{Продолжение таблицы 1} \\\\\\\\\\\\hline
    a & b & c & d\\\\\\\\ \\\\hline

    \\\\endhead
    \\\\endfoot
    \\\\endlastfoot

1 & 2 & 3 & 4\\\\\\\\ \\\\hline
t & r & e & z\\\\\\\\ \\\\hline

\\\\end{longtable}
\\\\fontsize{\\\\defaultfontsize}{\\\\defaultfontsize}\\\\selectfont\\\\setstretch{1.5}
"
`;

exports[`simple md to latex docs printer bold and italic 1`] = `
"\\\\textbf{Bold}: \\\\textit{testing}
"
`;

exports[`url variants Bold url 1`] = `
"\\\\textbf{https://example.com/index\\\\_page.html?asd=asdasd\\\\&gege=gegege\\\\#header}
"
`;

exports[`url variants Default url 1`] = `
"\\\\underline{https://example.com/index\\\\_page.html?asd=asdasd\\\\&gege=gegege\\\\#header}
"
`;

exports[`url variants Italic url 1`] = `
"\\\\textit{https://example.com/index\\\\_page.html?asd=asdasd\\\\&gege=gegege\\\\#header}
"
`;

exports[`url variants No escape & code url 1`] = `
"\\\\texttt{https://example.com/index_page.html?asd=asdasd&asdasd=gege#header}
"
`;

exports[`url variants Underlined url 1`] = `
"\\\\underline{https://example.com/index\\\\_page.html?asd=asdasd\\\\&gege=gegege\\\\#header}
"
`;
